File: backend-go/internal/handlers/admin/canonical.go
package admin

import (
	"encoding/json"
	"net/http"

	"github.com/jmoiron/sqlx"
	"github.com/lib/pq"
)

// CanonicalGroupRow representa um canonical_product com contagem de filhos e marketplaces cobertos.
type CanonicalGroupRow struct {
	ID             int64          `db:"id"              json:"id"`
	TitleCanonical string         `db:"title_canonical" json:"title_canonical"`
	BrandID        *int64         `db:"brand_id"        json:"brand_id,omitempty"`
	PriceBand      *string        `db:"price_band"      json:"price_band,omitempty"`
	LowConfidence  bool           `db:"low_confidence"  json:"low_confidence"`
	CreatedAt      string         `db:"created_at"      json:"created_at"`
	ChildrenCount  int            `db:"children_count"  json:"children_count"`
	Marketplaces   pq.StringArray `db:"marketplaces"    json:"marketplaces"`
}

// fetchCanonicalGroups busca canonical_products com contagem de filhos e marketplaces.
func fetchCanonicalGroups(r *http.Request, db *sqlx.DB) ([]CanonicalGroupRow, error) {
	var rows []CanonicalGroupRow
	err := db.SelectContext(r.Context(), &rows, `
		SELECT
			cp.id,
			cp.title_canonical,
			cp.brand_id,
			cp.price_band,
			cp.low_confidence,
			cp.created_at::text AS created_at,
			COUNT(c.id) AS children_count,
			ARRAY_AGG(DISTINCT split_part(c.dedup_key, ':', 1)) FILTER (WHERE c.dedup_key IS NOT NULL) AS marketplaces
		FROM canonical_products cp
		LEFT JOIN catalog c ON c.canonical_product_id = cp.id
		GROUP BY cp.id
		ORDER BY cp.created_at DESC
		LIMIT 100
	`)
	return rows, err
}

// CanonicalChildRow — item de catálogo vinculado a um canonical_product (filho do grupo).
type CanonicalChildRow struct {
	ID           int64    `db:"id"            json:"id"`
	Title        string   `db:"title"         json:"title"`
	SourceID     string   `db:"source_id"     json:"source_id"`
	Marketplace  string   `db:"marketplace"   json:"marketplace"`
	PriceCurrent *float64 `db:"price_current" json:"price_current,omitempty"`
}

// ListCanonicalChildrenHandler implementa GET /api/admin/canonical-groups/{id}/children.
// Lista os itens de catálogo (catalog) vinculados a um canonical_product via
// canonical_product_id — os "filhos" exibidos ao expandir um grupo canônico.
func ListCanonicalChildrenHandler(db *sqlx.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, ok := pathInt(r, "id")
		if !ok {
			writeErr(w, http.StatusBadRequest, "id inválido")
			return
		}
		var rows []CanonicalChildRow
		err := db.SelectContext(r.Context(), &rows, `
			SELECT c.id, c.title, c.source_id,
			       split_part(c.dedup_key, ':', 1) AS marketplace,
			       c.price_current
			FROM catalog c
			WHERE c.canonical_product_id = $1
			ORDER BY c.price_current ASC NULLS LAST
			LIMIT 200
		`, id)
		if err != nil {
			writeErr(w, http.StatusInternalServerError, "erro ao buscar filhos: "+err.Error())
			return
		}
		if rows == nil {
			rows = []CanonicalChildRow{}
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(rows)
	}
}

// ListCanonicalGroupsHandler implementa GET /api/admin/canonical-groups.
// Retorna canonical_products com contagem de filhos catalog e marketplaces cobertos.
// Tabela vazia retorna [] sem 500.
func ListCanonicalGroupsHandler(db *sqlx.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		rows, err := fetchCanonicalGroups(r, db)
		if err != nil {
			writeErr(w, http.StatusInternalServerError, "erro ao buscar canonical groups: "+err.Error())
			return
		}
		if rows == nil {
			rows = []CanonicalGroupRow{}
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(rows)
	}
}

File: frontend/src/lib/api/canonical.ts
import { authFetch } from '../authFetch'

// CanonicalProduct representa um produto canônico consolidado entre marketplaces.
// A lista (/canonical-groups) traz contagem e marketplaces; os filhos são
// buscados sob demanda ao expandir (fetchCanonicalGroupChildren).
export interface CanonicalProduct {
  id: number
  name: string
  low_confidence: boolean
  marketplaces: string[]
  marketplace_count: number
  children_count: number
}

// CanonicalChild representa um item de catálogo filho vinculado ao produto canônico.
export interface CanonicalChild {
  id: number
  title: string
  source_id: string
  marketplace: string
  price_current?: number
}

// CanonicalGroupsFilter contém os filtros suportados pela listagem de grupos canônicos.
export interface CanonicalGroupsFilter {
  brand?: string
  low_confidence?: boolean
}

// Linha crua do backend GET /api/admin/canonical-groups.
interface CanonicalGroupFlat {
  id: number
  title_canonical: string
  brand_id: number | null
  low_confidence: boolean
  children_count: number
  marketplaces: string[] | null
}

// fetchCanonicalGroups busca a lista de grupos canônicos com filtros opcionais.
// Mapeia a resposta crua do backend (title_canonical, marketplaces[]) para o
// shape da UI. Retorna [] se o endpoint retornar 404 (feature não habilitada).
export async function fetchCanonicalGroups(
  filter: CanonicalGroupsFilter = {},
): Promise<CanonicalProduct[]> {
  const params = new URLSearchParams()
  if (filter.brand) {
    params.set('brand', filter.brand)
  }
  if (filter.low_confidence !== undefined) {
    params.set('low_confidence', String(filter.low_confidence))
  }

  const query = params.toString()
  const url = query ? `/api/admin/canonical-groups?${query}` : '/api/admin/canonical-groups'
  const r = await authFetch(url)

  if (r.status === 404) {
    return []
  }
  if (!r.ok) {
    throw new Error(`fetchCanonicalGroups ${r.status}`)
  }
  const raw = (await r.json()) as CanonicalGroupFlat[]
  if (!Array.isArray(raw)) return []
  return raw.map(g => {
    const marketplaces = Array.isArray(g.marketplaces) ? g.marketplaces.filter(Boolean) : []
    return {
      id: g.id,
      name: g.title_canonical || `#${g.id}`,
      low_confidence: !!g.low_confidence,
      marketplaces,
      marketplace_count: marketplaces.length,
      children_count: g.children_count ?? 0,
    }
  })
}

// fetchCanonicalGroupChildren busca os itens de catálogo filhos de um grupo canônico
// (sob demanda, ao expandir). Retorna [] em 404.
export async function fetchCanonicalGroupChildren(id: number): Promise<CanonicalChild[]> {
  const r = await authFetch(`/api/admin/canonical-groups/${id}/children`)
  if (r.status === 404) {
    return []
  }
  if (!r.ok) {
    throw new Error(`fetchCanonicalGroupChildren ${r.status}`)
  }
  const raw = await r.json()
  return Array.isArray(raw) ? (raw as CanonicalChild[]) : []
}
